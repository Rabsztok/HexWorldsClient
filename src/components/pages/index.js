import WorldsPage from './worlds_page'
import WorldPage from './world_page'
import NotFoundPage from './not_found_page'

export { WorldsPage, WorldPage, NotFoundPage }
