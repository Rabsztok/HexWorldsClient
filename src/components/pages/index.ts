import WorldsPage from './worlds_page'
import WorldPage from './world_page'
import NotFoundPage from './not_found_page'
import DebugPage from './debug_page'

export { WorldsPage, WorldPage, NotFoundPage, DebugPage }
