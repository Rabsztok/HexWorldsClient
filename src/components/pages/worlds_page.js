import React from 'react'
import { Link } from 'react-router-dom'
import { observer, inject } from 'mobx-react'
import {
  Card,
  CardContent,
  Grid,
  Typography,
  IconButton
} from '@material-ui/core'
import DeleteIcon from '@material-ui/icons/Delete'
import EnlargeIcon from '@material-ui/icons/ZoomOutMap'
import styles from 'styles/pages/worlds_page.module.scss'
import BottomNavigation from 'components/bottom_navigation'
import WorldDialog from 'components/world_dialog'

const WorldsPage = ({
  store: {
    worldStore: { worldsList, expand, remove }
  }
}) => (
  <div className={styles.list}>
    <Grid container justify="center" spacing={16}>
      {worldsList.map(world => (
        <Grid key={world.id} item xs={12} sm={8}>
          <Card className={styles.item}>
            <CardContent>
              <Typography type="title">
                <Link
                  to={world.ready ? `/world/${world.id}` : '#'}
                  className={styles.link}
                >
                  {world.name}
                </Link>
              </Typography>

              <Typography type="body1">
                <span>state: {world.state}</span>
                {' / '}
                <span>regions: {world.regions.length}</span>
              </Typography>

              <div className={styles.actions}>
                <IconButton
                  onClick={() => expand(world.id)}
                  disabled={!world.ready}
                >
                  <EnlargeIcon />
                </IconButton>
                <IconButton onClick={() => remove(world.id)}>
                  <DeleteIcon />
                </IconButton>
              </div>
            </CardContent>
          </Card>
        </Grid>
      ))}
    </Grid>

    <BottomNavigation>
      <WorldDialog />
    </BottomNavigation>
  </div>
)

export { WorldsPage }
export default inject('store')(observer(WorldsPage))
