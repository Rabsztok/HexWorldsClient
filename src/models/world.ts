import { countBy } from 'lodash'
import { types, Instance } from 'mobx-state-tree'
import Tile, { ITile } from 'models/tile'
import Region, { IRegion } from 'models/region'

const World = types
  .model('World', {
    id: types.identifier,
    size: types.number,
    name: types.string,
    regions: types.map(Region),
    tileMatrix: types.map(types.reference(types.late(() => Tile)))
  })
  .views(self => ({
    get regionsList(): IRegion[] {
      return Array.from(self.regions.values())
    },
    get state(): string {
      const regionsState = countBy(this.regionsList, 'state')
      console.log(regionsState)
      if (regionsState.in_progress > 0) return 'generating regions'
      if (regionsState.empty > 0) return 'populating map'
      return 'ready'
    },
    get ready(): boolean {
      return this.state === 'ready'
    },
    findTile(x: number, y: number, z: number): ITile | undefined {
      return self.tileMatrix.get([x, y, z].join(','))
    }
  }))
  .actions(self => ({
    addTile(tile: ITile) {
      const { x, y, z } = tile
      self.tileMatrix.set([x, y, z].join(','), tile.id)
    },
    addRegion(region: IRegion) {
      self.regions.set(region.id, region)
    },
    update({ regions, ...attributes }: { regions: any }) {
      console.log('update', regions, attributes)
      Object.assign(self, attributes)
      if (regions)
        regions.forEach((region: IRegion) =>
          self.regions.set(region.id, region)
        )
    }
  }))

export interface IWorld extends Instance<typeof World> {}

export default World
