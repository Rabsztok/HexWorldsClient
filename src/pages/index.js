import WorldsPage from './worlds_page/worlds_page'
import WorldPage from './world_page/world_page'
import NotFoundPage from './not_found_page'

export { WorldsPage, WorldPage, NotFoundPage }
