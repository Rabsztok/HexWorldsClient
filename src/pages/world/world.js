import React, {Component} from 'react'
import {observer} from 'mobx-react'
import tileStore from 'stores/tile_store'
import worldStore from 'stores/world_store'
import Canvas from 'components/canvas'
import Menu from 'components/canvas_menu/canvas_menu'
import styles from './world.scss'

@observer
export default class WorldPage extends Component {
  componentWillMount() {
    const world = worldStore.find(this.props.match.params.id)
    worldStore.selectWorld(world)
  }

  render() {
    if (worldStore.currentWorld) {
      return (
          <div>
            <Menu/>

            {tileStore.loading && <div className={styles.loading}><i className='fa fa-circle-o-notch fa-spin'/></div>}

            <div className={styles.container}>
              <Canvas/>
            </div>
          </div>
      )
    }
    else return null
  }
}
