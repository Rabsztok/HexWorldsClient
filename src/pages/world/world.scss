.container {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  overflow: hidden;
}

.loading {
  position: absolute;
  width: 100%;
  height: 100%;
  top: 0;
  left: 0;
  z-index: 999;
  color: white;
  font-size: 30px;

  i {
    position: absolute;
    bottom: 10px;
    right: 10px;
  }
}