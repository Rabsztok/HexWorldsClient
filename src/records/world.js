import { computed, action, observable } from 'mobx'
import { countBy } from 'lodash'

class World {
  @observable regions = []

  constructor(world) {
    Object.assign(this, world)
  }

  @action
  update(attributes) {
    Object.assign(this, attributes)
  }

  @computed get regionState() {
    return countBy(this.regions, 'state')
  }

  @computed get state() {
    if (this.regionState.in_progress > 0) return 'generating regions'
    if (this.regionState.empty > 0) return 'populating map'
    return 'ready'
  }

  @computed get ready() {
    return this.state === 'ready'
  }
}

export default World