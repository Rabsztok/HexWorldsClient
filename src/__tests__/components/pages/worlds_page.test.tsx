import React from 'react'
import { shallow } from 'enzyme'
import { WorldsPage } from 'components/pages/worlds_page'
import world from '__mocks__/world.mock.js'

const noop = () => {}

describe('WorldsPage', () => {
  let store: any

  beforeEach(() => {
    store = {
      worldStore: {
        worldsList: [world],
        expand: noop,
        remove: noop
      }
    }
  })

  it('renders without errors', () => {
    const component = shallow(<WorldsPage store={store} />)
    expect(component).toMatchSnapshot()
  })
})
