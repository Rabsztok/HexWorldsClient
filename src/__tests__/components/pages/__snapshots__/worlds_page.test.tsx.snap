// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorldsPage renders without errors 1`] = `ShallowWrapper {}`;
